## ``Classification Models - Loan Default``

##### ``Sandra Ranilla-Cortina``

In [1]:
# Libraries
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

plt.style.use('fivethirtyeight')
np.random.seed(777)

import warnings
warnings.filterwarnings("ignore", category = UserWarning)
warnings.filterwarnings("ignore", category = FutureWarning)

#### ``Loan Default Pre-Processed Data + Dimensionality Reduction (PCA)``

In [2]:
# Loan Default Pre-Processed data + PCA (as LoanDefault does not have lots of attributes -> better try with the entire dataset)
X_train = pd.read_pickle('https://raw.githubusercontent.com/sranillac/SRC_UniOvi/refs/heads/main/ML-IA/data/LoanDefault/X_Train_LoanDefault.pkl')
X_test = pd.read_pickle('https://raw.githubusercontent.com/sranillac/SRC_UniOvi/refs/heads/main/ML-IA/data/LoanDefault/X_Test_LoanDefault.pkl')
Y_train = pd.read_pickle('https://raw.githubusercontent.com/sranillac/SRC_UniOvi/refs/heads/main/ML-IA/data/LoanDefault/Y_Train_LoanDefault.pkl')
Y_test = pd.read_pickle('https://raw.githubusercontent.com/sranillac/SRC_UniOvi/refs/heads/main/ML-IA/data/LoanDefault/Y_Test_LoanDefault.pkl')
print(X_train.shape, X_test.shape)
print(Y_train.shape, Y_test.shape)

(191510, 16) (63837, 16)
(191510, 1) (63837, 1)


#### ``Logistic Regression``

In [3]:
# Logistic Regression Model (LR)
log_reg = LogisticRegression(max_iter = 1000, random_state = 42)
log_reg.fit(X_train, Y_train)

LogisticRegression(max_iter=1000, random_state=42)

In [4]:
# Prediction
Y_pred_LR = log_reg.predict(X_test)

#### ``Support Vector Machine (SVM)``

In [5]:
# Support Vector Machine Model (SVM)
svm_model = SVC(kernel = 'linear', random_state = 42)
svm_model.fit(X_train, Y_train)

SVC(kernel='linear', random_state=42)

In [6]:
# Prediction
Y_pred_SVM = svm_model.predict(X_test)

#### ``K-Neighbors``

In [7]:
# K-Neighors Model (KNN)
knn_model = KNeighborsClassifier(n_neighbors = 4)
knn_model.fit(X_train, Y_train)

KNeighborsClassifier(n_neighbors=4)

In [8]:
# Prediction
Y_pred_KNN = knn_model.predict(X_test)

#### ``Decision Tree``

In [9]:
# Decision Tree Model (DT)
dt_model = DecisionTreeClassifier(max_depth = 5, random_state = 42)
dt_model.fit(X_train, Y_train)

DecisionTreeClassifier(max_depth=5, random_state=42)

In [10]:
# Prediction
Y_pred_DT = dt_model.predict(X_test)

#### ``Random Forest``

In [11]:
# Random Forest Model (RF)
rf_model = RandomForestClassifier(n_estimators = 100, max_depth = 5, random_state = 42)
rf_model.fit(X_train, Y_train)

RandomForestClassifier(max_depth=5, random_state=42)

In [12]:
# Evaluation
Y_pred_RF = rf_model.predict(X_test)

#### ``Evaluation Metrics``

1) ``Confusion matrix`` is a table that allows us to visualize the performance of a classification model:  
    - True Positives (**TP**): number of instances correctly predicted as positive
    - True Negatives (**TN**): number of instances correctly predicted as negative
    - False Positives (**FP**): number of instances incorrectly predicted as positive
    - False Negatives (**FN**): number of instances incorrectly predicted as negative.  


2) ``Accuracy`` measures the proportion of correct predictions (both positive and negative) out of the total predictions made:  

    $\text{Accuracy} = \frac{TP + TN}{TP + TN + FP + FN}$  

3) ``Precision`` measures the proportion of positive predictions that are actually correct:  

    $\text{Precision} = \frac{TP}{TP + FP}$  

4) ``Recall`` measures the proportion of actual positive instances that were correctly identified by the model:  

     $\text{Recall} = \frac{TP}{TP + FN}$  

5) ``F1-Score`` is the harmonic mean between **Precision** and **Recall**. It is useful when there is an imbalance in the classes:

     $\text{F1-Score} = 2 \times \frac{\text{Precision} \times \text{Recall}}{\text{Precision} + \text{Recall}}$

In [13]:
# Evaluation Metrics function
def eval_metrics(y, y_pred):
    mc = confusion_matrix(y, y_pred)
    acc = accuracy_score(y, y_pred)
    cr = classification_report(y, y_pred)
    return mc, acc, cr

In [14]:
# Logistic Regression Model - Evaluation
mc_LR, acc_LR, cr_LR = eval_metrics(Y_test, Y_pred_LR)

print(f'Logistic Regression Model - Confusion Matrix:')
print(mc_LR)
print('')
print(f'Logistic Regression Model - Accuracy: {acc_LR:.2f}')
print('')
print(f'Logistic Regression Model - Classification Report:')
print(cr_LR)

Logistic Regression Model - Confusion Matrix:
[[56289   135]
 [ 7186   227]]

Logistic Regression Model - Accuracy: 0.89

Logistic Regression Model - Classification Report:
              precision    recall  f1-score   support

           0       0.89      1.00      0.94     56424
           1       0.63      0.03      0.06      7413

    accuracy                           0.89     63837
   macro avg       0.76      0.51      0.50     63837
weighted avg       0.86      0.89      0.84     63837



In [15]:
# SVM Model - Evaluation
mc_SVM, acc_SVM, cr_SVM = eval_metrics(Y_test, Y_pred_SVM)

print(f'SVM Model - Confusion Matrix:')
print(mc_SVM)
print('')
print(f'SVM Model - Accuracy: {acc_SVM:.2f}')
print('')
print(f'SVM Model - Classification Report:')
print(cr_SVM)

SVM Model - Confusion Matrix:
[[56424     0]
 [ 7413     0]]

SVM Model - Accuracy: 0.88

SVM Model - Classification Report:
              precision    recall  f1-score   support

           0       0.88      1.00      0.94     56424
           1       0.00      0.00      0.00      7413

    accuracy                           0.88     63837
   macro avg       0.44      0.50      0.47     63837
weighted avg       0.78      0.88      0.83     63837



In [16]:
# K-Neighbors Model - Evaluation
mc_KNN, acc_KNN, cr_KNN = eval_metrics(Y_test, Y_pred_KNN)

print(f'KNN Model - Confusion Matrix:')
print(mc_KNN)
print('')
print(f'KNN Model - Accuracy: {acc_KNN:.2f}')
print('')
print(f'KNN Model - Classification Report:')
print(cr_KNN)

KNN Model - Confusion Matrix:
[[56015   409]
 [ 7284   129]]

KNN Model - Accuracy: 0.88

KNN Model - Classification Report:
              precision    recall  f1-score   support

           0       0.88      0.99      0.94     56424
           1       0.24      0.02      0.03      7413

    accuracy                           0.88     63837
   macro avg       0.56      0.51      0.48     63837
weighted avg       0.81      0.88      0.83     63837



In [17]:
# Decision Tree Model - Evaluation
mc_DT, acc_DT, cr_DT = eval_metrics(Y_test, Y_pred_DT)

print(f'Decision Tree Model - Confusion Matrix:')
print(mc_DT)
print('')
print(f'Decision Tree Model - Accuracy: {acc_DT:.2f}')
print('')
print(f'Decision Tree Model - Classification Report:')
print(cr_DT)

Decision Tree Model - Confusion Matrix:
[[56255   169]
 [ 7197   216]]

Decision Tree Model - Accuracy: 0.88

Decision Tree Model - Classification Report:
              precision    recall  f1-score   support

           0       0.89      1.00      0.94     56424
           1       0.56      0.03      0.06      7413

    accuracy                           0.88     63837
   macro avg       0.72      0.51      0.50     63837
weighted avg       0.85      0.88      0.84     63837



In [18]:
# Random Forest Model - Evaluation
mc_RF, acc_RF, cr_RF = eval_metrics(Y_test, Y_pred_RF)

print(f'Random Forest Model - Confusion Matrix:')
print(mc_RF)
print('')
print(f'Random Forest Model - Accuracy: {acc_RF:.2f}')
print('')
print(f'Random Forest Model - Classification Report:')
print(cr_RF)

Random Forest Model - Confusion Matrix:
[[56424     0]
 [ 7413     0]]

Random Forest Model - Accuracy: 0.88

Random Forest Model - Classification Report:
              precision    recall  f1-score   support

           0       0.88      1.00      0.94     56424
           1       0.00      0.00      0.00      7413

    accuracy                           0.88     63837
   macro avg       0.44      0.50      0.47     63837
weighted avg       0.78      0.88      0.83     63837



Which is the **``best performing model``**?  

- The **Logistic Regression** model achieves $89\%$ accuracy, with a strong precision of $0.89$ for **class 0**. However, it struggles significantly with **class 1**, achieving a recall of just $0.03$, indicating poor performance in identifying **default** cases.  

- The **SVM** and **Random Forest** models both have an accuracy $88\%$ accuracy, but they completely fail to classify **class 1**, with a recall of $0$. These models show no ability to recognize the **default** class.  

- The **Decision Tree** model also has an accuracy of $88\%$ and performs somewhat better than SVM and Random Forest in identifying **class 1**, with a recall of $0.03$, though still very low.  

- The **KNN** model also has an accuracy of $88\%$, but it performs slightly better than the other models in recognizing **class 1**, with a recall of $0.02$, though still very low.  

Therefore, **``Logistic Regression``** model, despite its poor recall for **class 1**, is the best-performing model overall. Note that, the low recall for **class 1** indicates that it would significantly benefit from techniques such as class balancing, feature engineering, or model optimization to improve its detection of the **``default``** class. Additionally, exploring more advanced models like **XGBoost**, **LightGBM**, or **Neural Networks** could further improve performance.